### PR TITLE
[Orgs] Persist `slug`

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This (re-)introduces persistence of `Organization.slug`.

New organizations will be given unique slugs. In case of conflicts, the slugified name attribute will be appended by a random string.

Org owners will be able to change the slug under Settings.

Screenshot 2023-03-21 at 10 41 05

### Known follow-ups:
 * backfill slugs for relevant entries
 * display `slug` for non-owner of the organization
 * add `slug` field to New Organization screen

## Related Issue(s)

Part of #16856

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by AlexTugarev** [comment]
Just found another issue I'd like to tackle before setting to ready: when changing slug, it needs to get validated.

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:179]: **thought**: It's feels right we are going down this route, see relevant discussions[[1](https://gitpod.slack.com/archives/C04E8MU39JQ/p1676047161150339?thread_ts=1676036280.889149&cid=C04E8MU39JQ)][[2](https://gitpod.slack.com/archives/C021Q4BL8E8/p1625064176314500?thread_ts=1625063886.313200&cid=C021Q4BL8E8)]. 💭

Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:179]: **suggestion**: If possible, I'd show only 8 characters here to make this more readable. Thoughts?

| BEFORE | AFTER |
|-|-|
| slug-before | slug-after |

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (gtsiolis) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:178]: **question**: Do we want to support slugs based on the org name when it's the first org and skip the suffix? It feels like we're rewarding the users grabbing a namespace first but is this something we want? Cc @atduarte

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamSettings.tsx:150]: **issue**: Using the same slug twice does not throw an error—we could probably not allow same slugs within an instance, right?

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/teams/TeamSettings.tsx:144]: **praise**: Thanks @AlexTugarev for using the new form components! Thanks also @selfcontained for initially adding[[1](https://github.com/gitpod-io/gitpod/blob/0a5235987358b7a935f572d4a1ff82f7207291ed/components/dashboard/src/components/forms/InputField.tsx#LL28C21-L28C21)] the [`htmlFor`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLLabelElement/htmlFor) element.

**Review by gtsiolis — APPROVED** [review]
Left a couple of UX comments below. 🏓 

Thanks, @AlexTugarev! 🌮 🌮 

Approving to unblock merging, holding to address the missing duplicate slug error and for someone to take a closer look at the code changes.

/hold

over-here

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (atduarte) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:178]: Yes 😁 If we need to take it back later, we will

Comment (gtsiolis) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:178]: ok

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:179]: changing to 8. fine with me.

File: components/dashboard/src/teams/TeamSettings.tsx
Comment (AlexTugarev) [reply components/dashboard/src/teams/TeamSettings.tsx:150]: Thanks for catching this!

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:192]: This can race and allow duplicate slugs. Can we push the check for existing into a single `create` method which does this in a transaction?

Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:192]: Good point. Using a db transaction should be feasible. ⏳

Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts]: done

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:186]: I'm guessing we've ensured that the resulting slug from this function does actually conform to the regex which we're enforcing when doing the update.

**Review by easyCZ — APPROVED** [review]

File: components/gitpod-db/src/typeorm/team-db-impl.ts
Comment (geropl) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:160]: Would be cool if we could limit slug to be no longer than 63 bytes; that way we could directly use them as subdomain in the future.

But also, this is a future problem we can absolutely solved later. :wink:

Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:160]: I'm fine with 63 bytes/chars.

**Comment by AlexTugarev** [comment]
/hold cancel

We'll verify on gitpod-staging.com once more, as the preview env doesn't work anymore.

**Pull request merged**
